cluster/assemblestate/assemblestate.go

```diff
@@ -0,0 +1,725 @@
+package assemblestate
+
+import (
+	"context"
+	"crypto/hmac"
+	"crypto/sha512"
+	"crypto/tls"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"log/slog"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/randutil"
+)
+
+// Transport provides an abstraction for defining how incoming and outgoing
+// messages are handled in an assembly session.
+type Transport interface {
+	// Serve starts a server that handles incoming requests and routes them to
+	// the provided [AssembleState].
+	Serve(ctx context.Context, addr string, cert tls.Certificate, as *AssembleState) error
+
+	// NewClient creates a client for sending outbound messages compatible with
+	// this [Transport].
+	NewClient(cert tls.Certificate) Client
+
+	// Stats returns the sent and received byte counts for this assembly
+	// session.
+	Stats() (tx, rx int64)
+}
+
+// Client is used to communicate with our peers.
+type Client interface {
+	// Trusted sends a message to a trusted peer. Implementations must verify
+	// that the peer is using the given certificate.
+	Trusted(ctx context.Context, addr string, cert []byte, kind string, message any) error
+	// Untrusted sends a message to a peer that we do not yet trust. The
+	// certificate that the peer used to communicate is returned.
+	Untrusted(ctx context.Context, addr string, kind string, message any) (cert []byte, err error)
+}
+
+// Discoverer returns a set of addresses that should be considered for assembly.
+type Discoverer = func(context.Context) ([]string, error)
+
+// AssembleState contains this device's knowledge of the state of an assembly
+// session.
+type AssembleState struct {
+	st     *state.State
```

Comment:
can we make this state independent and make it a problem of the user of thsi to bridge things into state?